### DISCUSSION TITLE
What would be a good way to structure this Heterogenous Graph Forecasting problem?

### DISCUSSION THREAD

**Discussion by snknitin** [pyg-team/pytorch_geometric#7540] [Q&A] [answered]
Hello PyG Community , 

I'm working with a heterogenous graph with 2 node types , say  `A` and `B`. Ideally for a forecasting task, but instead of temporal snapshots or sequential graphs, this is one large static graph. Every `(a_1, b_1)` pair is associated with a weekly-average(across 52 weeks) value of the forecasted variable, `x` . We want to predict this value for new combinations of A and B, that we haven't seen before. Basically we want to leverage the inductive nature and make an inference. 

By this I mean, {b_1,b_2,b_3,....b_1000} with {a_1,a_2,a_3.....a_50} , have observed data `x` for some combinations , but if we introduce 
* new nodes of type B, {b_1232 , b_2245 , b_9999} that didn't exist previously ,  OR
* new connections  between existing B and A, that don't have an edge or data observed. 

So given these possibilities, 

1) Any ideas on what would be a good way to model this problem to build a dataset?   
2) Is this a link prediction problem? or some sort of probability ranking/recommendation for nodes A with every combination of B? 

Ideally we want to be able to predict or forecast these values for the upcoming year, based on the static graph 

Please let me know if you've tried something like this before. Looking for some ideas on how to structure the labeled dataset for this "forecasting" problem for the next year , given the current year average values.  Is there a better way to approach this, that comes to mind? I would love some help

**Comment by rusty1s** [comment]
Sorry for my late reply, this slipped under my radar. I think this is an edge regression problem: Given an unseen pair node `(a_1, b_1)`, forecast a series of regression values (52?). To help with that, the model can integrate forecasting results from neighbors from both `a_1` and `b_1` during the forward pass via a GNN.

**Reply by snknitin** [reply]
Thanks for your response.  I'm going down a similar line. I'm confused between keeping dynamic weekly snapshots and doing 52 graphs with a single dimension edge label or one static graph with a 52 dimension edge label.  The idea is to learn from 2-3 years of data  and predict on the next year, for 52 weeks. So if I do a static graph with a 52 dimension edge regression, how do I split the train , val , test, given that I want to predict the next year 52 week forecast ? 

`a1,b1` nodes and `edge_label` would be 52 dimensions , for say 2021. Then I need to predict for 2022 for both `a_new , b_new` but also `a1,b1` which should now have a different forecast for next year. 

OR

Do a weekly graph for 2 years and sequentially split the data into train, val, test.  But then for inference I have to create 52 weekly graphs with random edges and predict based on that.  

Wasn't sure which approach made more sense. Trying both

**Reply by rusty1s** [reply]
I think one static graph with a 52 dimension edge regression is easier to handle TBH. Best splitting depends on how much data you have. I would probably just use a random split on the last year of data, and use any information from earlier years for message passing inside the GNN.

**Reply by snknitin** [reply]
Oh, interesting.  let's say we have 3000 nodes of one type and 50000 nodes of the other. I do a random Link split to make batches? Also when you say use info from earlier year for message passing, you mean convert those to node features? that's how they get passed right?

Would you suggest,  I do a 52 dimension of current year as `edge_feature` and map it to next year's 52 dimension `edge_label` for predictions. Then I can pull in random combinations for both nodes to form edges, but the main issue is I wouldn't have any specific features to pull for edges of past year that didn't exist, but i want to predict them based on other existing edges . The objective is basically to learn from similar nodes of A and similar nodes of B and predict how b_x would perform in a_y, to do some sort of forecast and convert it to a ranking or recommendation problem. Given the sparsity and seasonality, i do expect some of the 52 dimensions would be 0

**Reply by rusty1s** [reply, marked as answer]
Let me try to clarify. From my understanding, you have a `edge_index` and a corresponding 52-dim feature vector `edge_attr`. Each edge belongs to a given year (e.g., `edge_time` vector). During training/evaluation, you make predictions for the last year `t`. As such, you can divide the set of edges into training/evaluation edges and message passing edges
```python
# The edges to train/evaluate against:
mask = edge_time == t
edge_label_index = edge_index[:, mask]
edge_label = edge_attr[mask]

# Use previous information within GNN:
edge_index = edge_index[:, ~mask]
edge_attr = edge_attr[~mask]
```
You can then simple split `edge_label_index` randomly into training, validation and test sets.

**Reply by snknitin** [reply]
oh, it's a bit different. I thought you meant use the 52 dimensions as `edge_label` for the regression. I like this concept of edge_time vectors. But what would be the label for the `edge_label_index` for predictions of last year during training. 

Let me rephrase my problem. 
* Given 2 nodes types, let's say  A (3000,250) and B (50000,430), I have some metric being tracked for 52 weeks for some (a,b) combinations.
*  I have that for last year `t0 ( 52 attr?)` and this year `t1( 52label?)`. 
* I want to be able to predict the 52 week forecast for `t2`, next year.  
* Not all the nodes in A and B are fully connected. Let's say we have maybe 1,250,000 edges (less than 50000 x 3000).  
* Now we want to try, say expanding the number of edges to some new possibilities and maybe introduce even new nodes to make it 4000 and 70000 and try other combinations along with the existing 1.25 million edges for getting the `52 weeks of t2` 

The inductive nature and weekly sparsity in some combinations, makes Graphs a very good alternative to xgboost/lstm forecast or predict for future and unseen combinations. I've had pretty good success with heterogenous GNNs . In this case, i'm wondering what would be a proper way to frame the `Graph dataset` and `Dataloader`. 

### Dynamic but weekly snapshots
Obviously, if I had a weekly graph dataset to make `(2*52) graphs`, i could do a `temporal split for train, val, test` and  i would make it a `single dimension regression problem`, but that would have scalability issues or a lot of `reconstruction and generation of 52 new graphs` for inference with all possible combinations that may predict a lot of 0's for some of the combinations and for parts of the 52 weeks. 

### Static Large graph
If I build `a large static graph with 1.25 million edges`, that already `encapsulate the 52 weeks of history and 52 weeks of ground truth in the attr and label`, a time based split won't work but i can `create all edge sampling combinations or add some new nodes` and create just one more graph and run inference on it. Batching would be easy but `similarity`(which is important for inductive) for message passing might be hard to take into account, if i take a `random node split or link split`.  I'll miss out on leveraging the whole graph for weekly forecasts 

So i'm trying to figure out how to let the model learn and predict out for 52 weeks. which will then be averaged out into a metric that helps rank better combinations of (a,b) including new unseen ones for recommendation. I'm leaning towards the static route while I keep doing some feature engineering for the nodes and edges right now.

**Reply by rusty1s** [reply]
Yes, I think that's how I understood the problem. My assumption is that you always use the previous year worth of data to make a prediction for the next year. So:

* Training/Evaluation: You train on `t0` data and use data from `t1` for supervision/evaluation/metrics
* Inference: You use the pre-trained model and apply it on data from `t1` to make a prediction for `t2`

**Reply by snknitin** [reply]
yes. I'll frame it this way and test it out. I'm also thinking of doing a sliding window approach to look at 52 weeks on either side of the timeline when building edges for each week's graph, to set that temporal aspect in the model. That might make it a bit complicated though

**Reply by qoffee** [reply]
This is interesting discussion. I have a similar task where I have about 5 variables with data for the previous 20 years (I want to predict the number of articles and the popularity of objects in these articles) and taking into account the current version of PyG library interfaces it is not very clear how to properly make splitters with a sliding window.
How did you ultimately solve your problem?

**Reply by snknitin** [reply]
For the sliding window, I did it before the graph construction. So basically the edges were multiple raw files. When i build the heterogrenous graph from raw files I collate them all as a GraphData object that has a set of graphs.  This is the same as the tutorial or example in the documentation for custom inMemoryData objects. It goes through the list of the files and makes one graph per each file. It isn't one huge graph. So batching is simple by indexing and taking 3-4 graphs at a time, and the edges were sliding sequences. So iteration is sequential. You'd use a regular split of these graphs into train, val and test by chronology sequence or depending on your objective
